### PR TITLE
Make it possible to write chunks larger than 4 GB

### PR DESCRIPTION
**Pull request by alexv1n**
Added a special chunk flag indicating that chunk size is a 64 bit value.
This allows to handle larger chunks (which heppens quite rarely) while
still maintaining backward compatibility with the majority of traces.

## Description
Here is my attempt at addressing chunk size limitations. See if I messed anything up.

### CODE REVIEW

File: renderdoc/serialise/serialiser.cpp
Comment (baldurk) [inline renderdoc/serialise/serialiser.cpp]: I'd rather check the flag and either write a 64-bit value or a 32-bit value, matching the logic in the read path. That way we don't make any unwanted endian-ness assumptions (even though I highly doubt that will ever change... you never know :smile: ).

File: renderdoc/serialise/serialiser.cpp
Comment (baldurk) [inline renderdoc/serialise/serialiser.cpp]: I find plain `0xFFFFFFFF` here more readable than `numeric_limits` etc. Unlike FLT_MAX or something it's immediately obvious what's being checked especially with the flag name.

**Review by baldurk — CHANGES_REQUESTED** [review]
I left a couple of minor comments, but I do think this is the best solution.

One other thing will be needed - at the moment there's nothing to stop an old version of RenderDoc from opening a file written after this change with the 64-bit chunk flag, which will then not be handled properly. Unfortunately the only thing we can do is bump the `SERIALISE_VERSION` value in `RDCFile` and write all .rdc files with a new version. Then in the loading code, we support both versions - there's already a pattern with `*InitParams::IsSupportedVersion` which could be added.

That way a new build of RenderDoc (after this change) would load old captures and new captures, and old builds can only open old captures. It's a big hammer to apply, but each individual API does something similar and forward compatibility is not something we ever plan to support, backwards compatibility is much more important.

Other than that I'll need to test this a bit myself before merging, I'll wait until the above is sorted.

File: renderdoc/serialise/serialiser.cpp
Comment (alexv1n) [reply renderdoc/serialise/serialiser.cpp]: Agreed

File: renderdoc/serialise/serialiser.cpp
Comment (alexv1n) [reply renderdoc/serialise/serialiser.cpp]: Done. I've been burnt by lots of ffff in constants like that (people writing 0xfffffff [missing one 'f']). C++14's digit separator does help ensure the proper values are typed (0xffff'ffff and 0x1'0000'0000).

Do you support C++14 at all yet? Or still have some platform that are restricted to C++11/0x?

**Comment by alexv1n** [comment]
> One other thing will be needed - at the moment there's nothing to stop an old version of RenderDoc from opening a file written after this change with the 64-bit chunk flag, which will then not be handled properly. Unfortunately the only thing we can do is bump the `SERIALISE_VERSION` value in `RDCFile` and write all .rdc files with a new version. Then in the loading code, we support both versions - there's already a pattern with `*InitParams::IsSupportedVersion` which could be added.

Yes, I was thinking of exactly the same thing. The only thing is that file header with version is written first, well before we know if there are any large chunks in the file. One solution would be to bump `SERIALISE_VERSION` as you suggested and write it always and if no large chunks encountered, and if it is possible to seek within the file, go back and fix up the version to v1.0... But if write is done in streamed mode, then leave it in the new version... Or start with the old version and upgrade if large chunks are encountered. But for streamed write, if the file does have large chunk, the trace will be corrupt. The decision here is to see what is more likely/problematic approach. It seems that chunk sizes >= 4GB are extremely rare. And if we use the second approach (write legacy and upgrade format version as needed), is the chance of producing corrupt file for streamed write is important enough, versus the fact that we'll be producing newer file format, which is incompatible with older versions, for 99.99% of cases were we shouldn't...

I guess it's your call. :)

File: renderdoc/serialise/serialiser.cpp
Comment (baldurk) [reply renderdoc/serialise/serialiser.cpp]: That's a fair concern, though hopefully something that can be caught in review. Where possible I might use `~0U` for something like this but my intuition is not to do that when mixing and matching between 32-bit and 64-bit since then it's quite confusing too.

The UI uses C++14 for a couple of tiny bits of code in the python bindings, but I only use C++11 in the core code. I haven't tested it but my assumption is that Android probably can't handle it because Android is completely terrible at everything.

**Comment by baldurk** [comment]
Yeh going down the road of trying to detect and dynamically set the version in either direction is going to be error prone and IMO not worth the effort. I know that people definitely want to be able to open older captures with new versions of RenderDoc so backwards compatibility is definitely important (one reason why we want to make this larger chunk-length a flag and why we don't just change the file format), but opening newer captures with older versions doesn't seem useful to me. If you have a new version to make the capture, then you can use it to open the capture as well as all your other captures.

It is a big breaking change for something that in practice won't come up in 99.99% of captures, but we are already pretty quick to break forwards compatibility in each API's serialisation - e.g. the serialisation of one struct/API call is changed so we bump the version number, without necessarily knowing if that struct/API call is actually ever going to be included. As above - it's just not worth the effort and potential risk of bugs to try and preserve forwards compatibility that is rarely (ever?) that important.

**Comment by alexv1n** [comment]
> it's just not worth the effort and potential risk of bugs to try and preserve forwards compatibility that is rarely (ever?) that important.

I'm thinking of one person capturing and another opening (using potentially different version). Like one team or company captures an .rdc file and then sends it to another team/company for analysis (e.g. to investigate a bug in driver, a game developer might capture the frame and send the .rdc file to driver folks to debug/fix the issue).

Bumping .rdc file format version will force the receiving end to upgrade their version of RenderDoc which may not always be possible in particular dev environment. So, trying to see if we can reduce the chances of forcing this as much as possible

**Comment by baldurk** [comment]
Yeh that's true but like I say that can also happen on a narrower scale per-API. On Vulkan the per-API version has bumped between all of v1.0, v1.1, v1.2, and v1.3/v1.4, so each of those versions is already not forwards compatible with the next.

I'm not really concerned about people with older versions, since RenderDoc requires no special permissions and can be built from source as needed. I'd expect anyone in RenderDoc's target audience to be able to use a newer version if that's what's required. Worst case in the situation you describe then the person who captured the frame can use an older version to capture instead if there's such a blocker.

Without doing the convoluted patching of versions then we don't know if the situation you describe would break anyway because the person capturing on a newer version needs 64-bit chunk sizes anyway.

**Comment by alexv1n** [comment]
Well, it's your decision to make. I'm just bringing up concerns to consider... :)

**Comment by alexv1n** [comment]
So, what are the next steps? Are you going to be doing some run-time testing? Should I be doing something with `SERIALISE_VERSION` in the meantime?

**Comment by baldurk** [comment]
Yes, please add handling to the `RDCFile` loading to bump the serialise version and load the old one silently. Then I'll do some testing once this PR is ready to merge.

**Comment by alexv1n** [comment]
Ok, so, here is RDC v1.1 handling. One question, should I bother with ignoring the 64 bit chunk flag if we are reading v1.0 files? Or can I just assume that that flag is just 0 and automatically get ignored (which is what happens right now). Or, I can make sure that we inspect the flag value only if we are reading v1.1 or greater file formats... Not sure if this is worth doing though...

**Comment by baldurk** [comment]
No that's fine, nothing would have been emitting unknown flags previously so that's as corrupt as if we read complete garbage. Hopefully the code shouldn't crash if it does encounter one incorrectly, but it's not something I'm that concerned about.

**Review by baldurk — APPROVED** [review]

**Comment by baldurk** [comment]
Mostly I'm going to take you at your word that this fixes the actual bug as I had difficulty properly testing that this change worked to fix the bug, but it doesn't regress as far as I could see. On Vulkan I couldn't get any driver to let me allocate enough memory, and on D3D12 I ran into a number of either driver bugs or API/runtime bugs I'm not sure.

I eventually got something loaded and found that there are a few places in the UI that don't properly handle 64-bit sizes or offsets - e.g. the buffer viewer uses Qt widgets for the offset and length which are based on 32-bit ints, so that will need to change. It's minor though so can be done in a follow-up commit.

**Pull request merged**
